### ISSUE TITLE
Headless Raspberry Pi support

### ISSUE THREAD

**Issue by jpkelly** [bitfocus/companion#313]
Starting Companion on boot of a Raspberry PI without the need for a display connected or logging in to a VNC or SSH session.

Preferably this would be enabled via a systemd service.

Some basic config options exposed to the command line such as setting the network interface

**Comment by willosof** [comment]
This is right around the corner, but it will for now require that you clone the repository and follow the developer guide, and not start it via the normal prebuilt packages.

**Comment by willosof** [comment]
Did you make this @haakonnessjoen ?

**Comment by haakonnessjoen** [comment]
There is a alpha version in the source repo: headless.js. But currently it isn't that straightforward to use for the end user, since you would need to install node, and run the update scripts before you can use it. So I think we need to do something clever before we close this, and also more testing.

**Comment by depili** [comment]
What would be the easiest way to import the configuration for the headless mode instance?

**Comment by haakonnessjoen** [comment]
On raspberry pi? Try cp ~/.config/companion/db ~/companion/ if I don't remember wrong.

The most user friendly way would be to start it normally, export the full config as file, and then run headless and import it again, using the gui.

**Comment by depili** [comment]
Getting the headless mode to run requires:
1. installing the dev environment, which has it's own issues on a fresh rasbian: #478 and #479 
2. running ./tools/build_writefile.sh
3. copying the configuration to ~/companion/
4. running ./headless.js wlan0 or ./headless.js eth0

**Comment by depili** [comment]
After all that I got the headless.js running. With osc data updates coming in every 30ms from clock-8001 the update rate on buttons with variables lags 1-4 seconds behind and after about 1,5 minutes something just gives up and the stream deck goes all black. 

In the all black state pressing a button on the stream deck causes it to return to the logo display after a small delay.

The companion code continues to run and still prints osc updates and button gfx generation messages into the console but everything is just dead on the stream deck side.

**Comment by depili** [comment]
Deleting all but the depili-clock-8001 module seems to improve the stability considerably. Now at over 6 minutes and counting where as before it never stayed stable for 2 minutes. The updates still lag behind and arrive in clumps with the display often skipping some second display updates.

**Comment by haakonnessjoen** [comment]
People have reported that installing the raspian lite image makes usb stable with stream deck using headless.js. Not sure about latency though.

**Comment by depili** [comment]
It seems that the memory needs are really marginal on the pi. After the module deletion and running only one module and 3 pages of buttons, of which 2 are updating from variables, one showing seconds and one minutes, the free memory on the pi hovers at around 70-65M running with rasbian lite.

**Comment by depili** [comment]
I don't like that when the updates cease and the stream deck goes black no error what so ever gets reported on the console nor does companion exit (with non-zero exit status preferably), so any kind of stability measurement or restart logic is impossible to implement.

**Comment by depili** [comment]
> 
> 
> People have reported that installing the raspian lite image makes usb stable with stream deck using headless.js. Not sure about latency though.

I have been using rasbian lite from the start.

**Comment by sphlabs** [comment]
I was having problems with the StreamDeck going black/disconnecting. It could be reconnected by rescanning USB from the surfaces tab.

I believe this was due an under-rated power supply to the Pi. The 3 B+ draws more power than previous models, added to the power draw needed by the Streamdeck, it is possible to overload smaller PSUs.

Check /var/log/kern.log for messages like `kernel: [ 1612.002574] Under-voltage detected! (0x000d0005)`

**Comment by depili** [comment]
Apparently the stability problem was indeed power related, one of my
trusty anker cables seems to be failing.

So will integrate a decent 5V 6A power supply into an enclosure with the
stream deck and the pi and feed the power in from the GPIO pins.

Still leaves something to be desired on the update rate of the stream deck
buttons. The rpi cpu is barely doing anything, so the bottleneck is
somewheere else, load average is at 0.52 at most for 1 minute average.

Memory usage also seems to be in control with only one companion module
included in the three, leaving most of the memnory free running on the
raspbian lite. Free memory seems to average about 570Mb.

**Comment by depili** [comment]
Testing with a 6A 5V power supply and a 4700uF smoothing capacitor
actually yields worse results, the stream deck goes away in under a minute
:/ no under voltage alerts in the log and no dmesg messages about usb
devices disconnecting or re-enumerating. Weird...

**Comment by sphlabs** [comment]
I've had good success so far, switching from a 1A to 2A USB power supply. Haven't done any extensive soak testing yet, but it seems to have fixed the problems for me.

**Comment by depili** [comment]
Will need to bring my laptop next time to troubleshoot further and verify
that the unit is working with the macbook, starting to think that it might
have a fault in the cable or something like that.

Tried to give 5,2V to the 5V rail and it is still unstable for the stream
deck, the web interface works just fine and rescanning re-aquires the
stream deck just fine until it disappears again. (no kernel messages about
renegotiating the usb link or anything like that.

**Comment by depili** [comment]
Continued testing with the latest head 9a404b03e3a15ea52c00c706511b30ba94916d4e button updates are steady but the stream deck still disconnects from companion after a short while. I have two different stream decks to test with currently and both are showing the same behaviour.

Running with raspberry pi 3b+ and stretch lite. Next up: testing with rpi 3b

**Comment by depili** [comment]
And same results with rpi3b :/

**Comment by canoemoose** [comment]
Depili - I'm trying this too, albeit using Irisdown Countdown Timer 2, which only provides feedback every second. My circumstances for disconnection are as follows:

- Running both headless and non-headless, and
- Displaying a page of buttons with feedback variables in their text - in this case the countdown time remaining

 I'm using a Pi 3b (non +)
I get the following when the stream deck goes blank:
`lib/elgato_dm remove device 0001:0014:00 +2m
  lib/device unloading for 0001:0014:00 +2m
  lib/action adding action bitfocus-companion:instance_control +2m
  lib/action adding action bitfocus-companion:set_page +1ms
  lib/action adding action bitfocus-companion:inc_page +1ms
  lib/action adding action bitfocus-companion:dec_page +0ms
  lib/action adding action bitfocus-companion:button_pressrelease +1ms
  lib/action adding action bitfocus-companion:button_press +0ms
  lib/action adding action bitfocus-companion:button_release +1ms
  lib/action adding action bitfocus-companion:panic +6ms
  lib/usb/elgato elgato.prototype.clearDeck() +2m
`
Rescanning in control panel finds it again.

Is this what you see?
The same config, running on the same Windows bleeding edge build does not exhibit this bug.

**Comment by depili** [comment]
Pretty similiar, don't have the exact logs. My gut tells me this might be
a timing issue, where the graphics generation takes enought time that it
blocks some usb transactions that end up timing out because of it.

This is plausible since nodejs executes code in one event loop per process
and if processing one event is ongoing all others are blocked until that
event finishes. The Pi being quite bit slower means the event loop is
blocked for longer periods of time and the likelihood of timing collisions
because of that increases greatly.

The companion code itself seems to be smart enough that it only updates
the graphics when they actually change and not on each feedback update (if
the data stays the same), which helps as my current implementation sends
the osc feedback on each frame.

Will need to test this by having a setup with no feedback at all, and thus
no button graphics regeneration, and see if it alters the stability. I
should also have access to more powerful bananapi board soonish to test.

If it indeed is caused by the event loop timing issues I don't know what
could be done other than trying to separate the gfx generation and usb
into diffirent nodejs processes. I haven't even tried to look into the
companion codebase to find out what kind of multiprocessing, if any, it
already does.

**Comment by canoemoose** [comment]
With regards to your further testing comments - if the Stream Deck is
showing a page without feedback (ie buttons are on Page 1, Stream Deck is
showng Page 2) then it's stable.
Likewise if the same buttons are shown without making use of the feedback
text.
(I had 3 buttons showing remaining hours, minutes and seconds of my
countdown timer)

**Comment by alatteri** [comment]
I'm having the button blacking out issue too.  Raspberry 3B+, fresh install of Lite, latest git head, Official wall wart PSU.

**Comment by Towerful** [comment]
I have encountered this problem. RPI3b+ running debian stretch.  
It seems to be an issue with the redrawing.  
If you SSH into it and run `top`. Then spam the page change and action buttons, you will see a node process start eating up resources, then get killed when the kernel (I think) is getting resource-starved.  
Companion is still running, and can still be accessed via the webUI.

I haven't been able to find the correct log/command to find WHY the process was killed.  
And I haven't been able to trace it back to the actual JS files that are run in that process.  
I'm not brilliant with linux!

It seems that redrawing the streamdeck takes a lot of CPU. Or it is really inefficient on RPi architecture.  
Luckily it is in a different thread, so it doesn't tank the whole program when it is killed.

There might be something that can be configured in the OS to stop it from killing the process - again, I dont know linux that well.    
Or there might be a better way to redraw the streamdeck. But I don't know enough about this codebase to even start suggesting possibilities!  
But when I get time, I will be poking around and try to reproduce it in a way I can then trace the issue

**Comment by Towerful** [comment]
@canoemoose I agree with starting a new issue for this specific problem.  
If I get time, I will get the latest commit installed on my RPi, and get to testing.  
for now, I'll leave this here:

-----------------------------

I have encountered this problem. RPI3b+ running debian stretch.  
It seems to be an issue with the redrawing.  
If you SSH into it and run `top`. Then spam the page change and action buttons, you will see a node process start eating up resources, then get killed when the kernel (I think) is getting resource-starved.  
Companion is still running, and can still be accessed via the webUI.

I haven't been able to find the correct log/command to find WHY the process was killed.  
And I haven't been able to trace it back to the actual JS files that are run in that process.  
I'm not brilliant with linux!

It seems that redrawing the streamdeck takes a lot of CPU. Or it is really inefficient on RPi architecture.  
Luckily it is in a different thread, so it doesn't tank the whole program when it is killed.

There might be something that can be configured in the OS to stop it from killing the process - again, I dont know linux that well.    
Or there might be a better way to redraw the streamdeck. But I don't know enough about this codebase to even start suggesting possibilities!  
But when I get time, I will be poking around and try to reproduce it in a way I can then trace the issue

**Comment by depili** [comment]
Actually it isn't the os killing the process (you would see a log entry and the pi isn't running out of resources) but the fact that the USB communication and picture generation all run inside the single threaded nodejs event loop.

If the image generation takes too much time (as happens with the rpi regularly) the usb communication handled by the same event loop times out and the connection to the stream deck drops on the level of the nodejs driver, the kernel itself doesn't see the streamdeck dropping.

The real solution would be rewriting companion such that picture generation and preferably the streamdeck handling would be in their own distinct processes and would communicate over sockets between the processes. The current setup can't take advantage of more than one core out of the four that a rpi has.

**Comment by depili** [comment]
A band-aid solution might be replacing the pngjs library used by companion with something that is able to use external threads / native utilities for the image generation, as doing computation inside the nodejs event loop is just asking for problems. This would probably cause great headaches for cross-platform compatibility unless you package utilities like imagemagick with the distribution binaries of companion.

**Comment by Towerful** [comment]
So, perhaps the images could be generated and cached when a button is created?  
Or - by image generation - do you mean placing the images in the correct place for each button?

**Comment by haakonnessjoen** [comment]
@depili where are you getting your information from?
For starters, each usb device is forked into a seperate process. It is not all done in the same runloop. And next off, nothing in the usb part is timing out upon sending data. There is no timeout. The deck can however time out and reboot if you push a button, and don’t read it from the device in a few seconds. But the hid driver in node is constantly polling for key press data. As for image creation. They are only generated when they change. All buttons are cached at all times. Changing a page on a deck simply transfers cached image data from ram (via ipc and then usb) to the device.

I have yet to find the real reason some stream decks disconnect from raspberry pi. But as far as I know, this only happens on rpi. No other similar embedded platforms (again: as far as I know)

But @depili, please keep a clear line between facts and suspicions (in lack of better words).

**Comment by jarodwsams** [comment]
@haakonnessjoen you mention that this doesn't seem to be an issue, that you know of, with other embedded platforms. Do you have one in particular that performed well? I'm trying to create a fully-enclosed "shot box" for my church that we can use in our smaller spaces rather than having to buy $$$$ control equipment (i.e. ATEM consoles). I picked up a Pi because I'm familiar with the platform, but if there's another similar platform that you know works, I'd love to give it a try.

**Comment by depili** [comment]
The banana pi M3 works somewhat, that board has really bad wifi and
sd-card performance and only supports one version of ubuntu with the
Chinease kernel...


I'm basing my hypothesis about the timing out on that the disconnects with
the RPI only happen with buttons with feedback active that update their
graphics without user interaction and the deck disappearing in quite
consistent time without any user interaction or power issues. Also all of
the work is done just by one of the pi's CPUs, there is no memory stress,
there are no voltage issues even when monitoring by a scope so the issue
really can't be anything else than somewhat timing related to the image
generation via feedbacks and some interaction of that with the usb
communication.

The PI is unique in that way that it has _really_ bad memory IO bandwith
among all of the clones, and most of the clones have also faster execution
in a single core.

The issue is really really consistent to trigger with just one button that
has a feedback variable updating that button once per second.

**Comment by haakonnessjoen** [comment]
All of the work is *not* done by one of the cores. That is just false.

@jarodwsams, I haven't had much time to test different devics, but Intel Sticks seems to work pretty good. Just make sure you have either 64bit windows, or linux installed on it. I think a lot of them come preinstalled with 32bit windows or something(?).

We are working on a program written in C (for speed) for the RPI (and other embedded devices), where the pi is only pushing usb image-data to the device, sort of acting like a usb extender over network, for companion. But we have to get companion 2.0 out, and a bunch of other stuff too, before we can release this. It also needs to be upgraded to support the XL. Since this project started a while ago.

**Comment by jarodwsams** [comment]
> We are working on a program written in C (for speed) for the RPI (and other embedded devices), where the pi is only pushing usb image-data to the device, sort of acting like a usb extender over network, for companion. But we have to get companion 2.0 out, and a bunch of other stuff too, before we can release this. It also needs to be upgraded to support the XL. Since this project started a while ago.

If you're referring to Companion Satellite, Viker was telling me about that on Slack yesterday and I got very excited.

**Comment by depili** [comment]
Looking further into the rpi issue; when the disconnect happens, something
triggers the elgato_dm to remove the device, the trigger doesn't come from
the usb process error handler. And at the point where the removal is
triggered the communication to the streamdeck still works at least enough
for the clear_deck command to work.

After making a hack by also commenting out the remove request handler
binding from elgato_dm.js the stream deck seems to be stable on the rpi.
Additional grepping and hunting will be necesserry to understand what
triggers the removal request and why.

**Comment by haakonnessjoen** [comment]
Great stuff! Sounds like you are getting close to figuring out the problem, also sounds like I will be really «redfaced» when/if you find out, based on your currfnt findings 😅

**Comment by Towerful** [comment]
This is great work! I don't think I have anything useful here. Juts trying to figure out how I can help!

If you are saying that the removal request is not coming from the usb handling process ( https://github.com/bitfocus/companion/blob/1929553e6447cd038f2c9329ad802481a61de3d5/lib/usb/elgato.js#L77 ), then I think it has to come from the draw function ( https://github.com/bitfocus/companion/blob/1929553e6447cd038f2c9329ad802481a61de3d5/lib/usb/elgato.js#L140 ).

Considering that this is only try/catching to draw, without any error logging, its probably worth putting an error report in here, and see what the fillImage method ( https://github.com/bitfocus/node-elgato-stream-deck/blob/25c9b17bc257c61580e27de5f977794850e7ff05/index.js#L180 ) is doing.  
Unless I am mis-understanding how the error handling is working!

**Comment by jarodwsams** [comment]
> Looking further into the rpi issue; when the disconnect happens, something triggers the elgato_dm to remove the device, the trigger doesn't come from the usb process error handler. And at the point where the removal is triggered the communication to the streamdeck still works at least enough for the clear_deck command to work. After making a hack by also commenting out the remove request handler binding from elgato_dm.js the stream deck seems to be stable on the rpi. Additional grepping and hunting will be necesserry to understand what triggers the removal request and why.

@depili What line(s) did you comment out in elgato_dm.js? I'd like to replicate your band-aid.

**Comment by depili** [comment]
And digging deeper the trigger for the disconnect is an exception on
drawing a tile on the deck, which is non-fatal exception that is just
being processed without a log message as a reason to disconnect the deck.

The exception trace on a write error looks like this:
   at StreamDeck.write
(/home/pi/src/companion/node_modules/elgato-stream-deck-clean/index.js:141:22)
    at StreamDeck._writePage1
(/home/pi/src/companion/node_modules/elgato-stream-deck-clean/index.js:264:15)
    at StreamDeck.fillImage
(/home/pi/src/companion/node_modules/elgato-stream-deck-clean/index.js:202:8)
    at elgato.draw (/home/pi/src/companion/lib/usb/elgato.js:140:19)
    at process.<anonymous>
(/home/pi/src/companion/lib/usb/handler.js:63:58)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emit (internal/child_process.js:762:12)
    at _combinedTickCallback (internal/process/next_tick.js:142:11)
    at process._tickCallback (internal/process/next_tick.js:181:9)


Adding a simple retry logic to the draw command like:
```
elgato.prototype.draw = function(key, buffer) {
        var self = this;
        var retry = true;
        buffer = self.handleBuffer(buffer);

        while(true) {
                try {
                        self.streamDeck.fillImage(self.reverseButton(key),
buffer);
                        return true;
                } catch (e) {
                        console.log(e);
                        if (retry == false) {
                                self.log('Error drawing:', e);
                                self.system.emit('elgatodm_remove_device',
self.devicepath);
                                return false;
                        }
                        retry = false;
                }
        }
}
```

Seems to be enough to keep the connection stable. The draw errors are
quite rare.

**Comment by depili** [comment]
A quick look by adding debug logging to all of the elgato_dm functions
didn't find any timing correlation for the drawing failures, but my quess
would be either there is something pre-installed on the raspian image that
tries to look for certain usb HID devices or something else than the
elgato_dm is doing usb scans that lock the devices briefly.

The failures seem to be so random and quite rare, with time between failed
writes being from 3 seconds to 5 minutes so trying to find what conflicts
with the access will be really annoying unless something just jumps out.

The issue not manifesting on the banana pi at all even after tens of hours
leads me to suspect that is some other package accessing the usb.

**Comment by Towerful** [comment]
perhaps a more elegant solution would look like...

```js
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
	        return true;
	} catch (e) {
                console.log(e)
	}

        // failed to redraw, try again
	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
	        return true;
	} catch (e) {
                self.log('Error drawing:', e);
                self.system.emit('elgatodm_remove_device', self.devicepath);
                return false;
	}
```

I would consider it a bit confusing to use a `while(true)` loop with `return` statements.  
However, I will leave it up to the maintainers to actually code it!

That's awesome work! If this gets merged, @depili, how can I buy you a beer?

**Comment by haakonnessjoen** [comment]
Great stuff guys!

It would be nice to see what the error actually is, when this happens on raspberry pi. If it is an exception that is differnet from when there would be an actual non-recoverable error. That way we could retry the drawing routine if it is in fact just a temporary problem, like it clearly is with the raspberry pi at times.

**Comment by jarodwsams** [comment]
@Towerful Trying your flavor now. I'll be able to test it later this evening. Quick ? because I'm not a coder but can follow instructions and replicate.

Does this last `return true;` get left alone, or taken out? I'm assuming it gets taken out because it was added explicitly to the first try().
![image](https://user-images.githubusercontent.com/22972834/59122441-62bca600-8920-11e9-989c-2a05634f71fe.png)

**Comment by depili** [comment]
Javascript unfortunately lacks the traditional retry keyword for
try-catch and the while loop is the "normal" way of implementing N
retries, in this case just having another retry seems to be enough.

My first quess was the usb-hid hotkey daemon rpi uses (triggerhappy) but
that didn't seem to be it. Catching what the conflict is would then need
luck with lsof polling or another hunch on where to look.

**Comment by depili** [comment]
The error probably comes from node-hid on the undelying layer from
https://github.com/node-hid/node-hid/blob/6f36b9fd5eae0ac05a4f0066730c33e2eab153fd/src/HID.cc#L182
(or in the similar location on the bundled version instead of GIT head.
And as such being certain if this is a recoverable case is quite hard, but
retries should be fast and given low number of permitted retries and then
giving up and erroring out I don't see why it would be a real problem.

**Comment by Towerful** [comment]
@jarodwsams 
the final return that your cursor is on is irrelevant.  
All code paths of the final try/catch statement have a return statement, so that final line will never be evaluated.  
No harm in taking it out, no harm in leaving it in.  

@depili fair enough regarding the retry count. The majority of the coding I've done has been for personal/internal stuff, so error handling has always been 'report it and abort it'. I've never really played with retrying stuff!  
You are definitely getting deeper into linux/node than I can really help with.

**Comment by haakonnessjoen** [comment]
But if we are retrying I think we should add a bit of delay at least. So wait 10 or 20 ms and retry. Or else I would think it would be very probable that it would get the same error again.

**Comment by depili** [comment]
Delay might be reasonable, but it will then block, so far just retrying
straight after seems to work, it all depends on the nature of the
conflict, a delay might make it work better, it might harm things
depending on the frequency of the other access...

We pretty much know that the conflict can repeat at least once in 3
seconds (shortest time between draw failures I have seen) so that is the
upper limit on how often the accesses happen.

Since currently the deck is updating one time every second or so
(depending on the rpi lag and load), I would say that the other access has
period of less than a second, so in this case having a delay in the retry
might be harmful.

**Comment by haakonnessjoen** [comment]
I am not talking about blocking delay. You don't do blocking delays in node. A setTimeout simply exits the function and runs the callback at a later time. Hence not blocking anything, but postponing the last write.

**Comment by Towerful** [comment]
Considering this isnt an async function, perhaps a `setTimeout` or a `nextTick` callback to try/retry the draw is in order

**Comment by haakonnessjoen** [comment]
Exactly what I was suggesting @Towerful. And this code is run in a separate process too, so if you would block anything, it would only block that one stream deck. But there is no blocking. :)

**Comment by depili** [comment]
The issue with that we don't really know what the frequency of the other
blocking access (or if it really is a other access and not the node-hid
library failing randomly, there have been few issues on that project like
it in the past) we can't know what the proper delay would be.

If the other access happens every second, then any delay less than a
second would be ok, if the other happens every 20ms and you use a 20ms
delay then you increase the risk of the retry hitting the same issue by
adding the delay.

The delay also decreases the response time, but considering the sluggish
updates on a rpi that doesn't seem like a concern.

It would require experimentation, but so far retrying once immediately
hasn't failed a single time.

**Comment by jarodwsams** [comment]
Something like this?

```javascript
elgato.prototype.draw = function(key, buffer) {
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
		return true;
	} catch (e) {
		console.log(e)
	}
	
	// failed to redraw, try again
	setTimeout(elgato.prototype.redraw, 20);
	}

elgato.prototype.redraw = function(key, buffer) {
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
		return true;
	} catch (e) {
		self.log('Error drawing:', e);
		self.system.emit('elgatodm_remove_device', self.devicepath);
		return false;
	}
}
```

**Comment by jarodwsams** [comment]
If that does what I think it does (again, not really a coder), it would theoretically be possible then to add a "success" variable and so long as success == false then keep retrying every 10 or 20 ms. Right?

**Comment by haakonnessjoen** [comment]
If you could, please try to keep your test running as long as possible, since this issue is highly unstable. So unless it never happens again, I think I will add a hybrid approach. Retry immediately, and then retry after x ms before giving up the third time.

@jarodwsams: I would have written that setTimeout as: setTimeout(self.redraw.bind(self), 20); So that the "self" in redraw points to the correct "this". In case you want to learn to be a node.js coder ;)

**Comment by depili** [comment]
A hybrid seems reasonable, and you definitely don't want to keep retrying
for ever since the write error could also be due to something as simple as
someone unplugging their streamdeck and then the retry loop would never
stop.

In case of potentially recoverable runtime errors you always want to limit
the number of retries.

**Comment by jarodwsams** [comment]
@haakonnessjoen so, like this, then?

```javascript
elgato.prototype.draw = function(key, buffer) {
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
		return true;
	} catch (e) {
		console.log(e)
	}
	
	// failed to redraw, try again
	setTimeout(self.redraw.bind(self), 20);
	}

elgato.prototype.redraw = function(key, buffer) {
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
		return true;
	} catch (e) {
		self.log('Error drawing:', e);
		self.system.emit('elgatodm_remove_device', self.devicepath);
		return false;
	}
}
```

**Comment by haakonnessjoen** [comment]
@jarodwsams: No that would always resend the data. The settimeout should be inside the catch

**Comment by Towerful** [comment]
If we are suggesting code for this...  

```js
        elgato.prototype.draw = function(key, buffer, attempts = 0) {
        attempts++;
        var self = this;
        buffer = self.handleBuffer(buffer);

        try {
                self.streamDeck.fillImage(self.reverseButton(key), buffer);
                return true;
        } catch (e) {
                if(attempts > 2) {
                        self.log('Error** drawing:', e);
                        self.system.emit('elgatodm_remove_device', self.devicepath);
                        return false;
                }
                setTimeout(self.draw(key, buffer, attempts).bind(self), 20) //not sure if the bind is required
                // alternatively a setImmediate() or nextTick()
        }
```

The issue is if the result of the `draw()` function is required.  
If the result (`true` or `false`) is relied upon in the execution elsewhere, then those implementations will need to be adapted to handle promises. Then to have the `draw()` method return a promise that is resolved immediately (upon success) or finally resolved/rejected by the timeout function.  
Or adapt the whole thing to be `async/await`

**Comment by jarodwsams** [comment]
@haakonnessjoen like so?
```javascript
elgato.prototype.draw = function(key, buffer) {
	var self = this;
	buffer = self.handleBuffer(buffer);

	try {
		self.streamDeck.fillImage(self.reverseButton(key), buffer);
		return true;
	} catch (e) {
		console.log(e)
		// failed to redraw, try again
		setTimeout(self.redraw.bind(self), 20);
		}
	
	}
```

**Comment by haakonnessjoen** [comment]
@Towerful your setTimeout would run the draw function immediately, not after 20 ms.

It should look like:
`setTimeout(self.draw.bind(self), 20, key, buffer, attempts)`

**Comment by Towerful** [comment]
Haha, so it does. fixed

**Comment by alatteri** [comment]
This is how I currently run my 3 Pi based Streamdecks.  Each one just shares USB using VirtualHere to a linux VM that runs Companion.  Nothing is running locally on the Pi's.


> We are working on a program written in C (for speed) for the RPI (and other embedded devices), where the pi is only pushing usb image-data to the device, sort of acting like a usb extender over network, for companion. But we have to get companion 2.0 out, and a bunch of other stuff too, before we can release this. It also needs to be upgraded to support the XL. Since this project started a while ago.

**Comment by jarodwsams** [comment]
@Towerful @haakonnessjoen
I had about a 45-minute test period this evening with the modified draw/redraw functions (without the retry delay & counter). Worked like a charm. I'll have another 5-hour testing period on Sunday morning and will report back in the afternoon. I think we've found a viable solution, even if we haven't yet found the root cause.

**Comment by jarodwsams** [comment]
I should add that the 45-minute testing tonight was done *without* a powered USB hub. The deck was connected directly to the RPi. I'll keep the powered USB hub out of the mix on Sunday to confirm viability.

**Comment by depili** [comment]
My test has been running over night now with the original code that
retries once and everything still works. Running on a rpi3b+ with the deck
connected directly on the rpi and with quite cheapo usb power adapter.

**Comment by depili** [comment]
Still working nicely after approx 24 hours with the original retry logic
(max one retry, no delay).

The same fix will need to be applied to the elgato original and mini
implementations more than likely, I only have the original model to test
with.

**Comment by jarodwsams** [comment]
@Towerful How's your fix flavor working, with the retry counter?

**Comment by Towerful** [comment]
@jarodwsams I haven't tested it. Was just trying to contribute, as I have an interest in this!   
I have some time later this week (or perhaps at the weekend) when I can hopefully tinker with this.  
I don't think I'll be able to contribute to the root cause of the issue, seems like it's too deep into the linux/node underbelly for me!

**Comment by depili** [comment]
I can't see that it would do anything worse than the single retry variant,
which has now been running for 48ish hours and even when forcing 10 tiles
to update by user interaction as fast as the pi can handle it I can't make
it fail in a ~minute of pressing the buttons.

I would consider this streamdeck disconnection issue solved with even the
most basic retry logic.

**Comment by jarodwsams** [comment]
Root cause is way too deep for me, too. I can drop your code into my RPi and test :) It looks like it should do exactly what we need it to, just in a slightly more elegant flavor.

**Comment by depili** [comment]
Probably the easiest way to hunt for the root cause would be to use
wireshark to capture the usb traffic, filter out the ethernet and wlan
from that and then try to deduce where the interference is coming from.

**Comment by jarodwsams** [comment]
My deck just dropped out 👎
I was making a lot of changes to several pages and triggering buttons over the previous minute or so.
I'll drop in @Towerful's code with the retry loop and try to replicate again.

**Comment by Towerful** [comment]
Does anyone have a reliable strategy for causing the drop out?  
Something like 2 pages with some buttons. A button on each page triggers a page change to the other page, then clicks a similar button to change page again? Perhaps with a slight delay?  
So it continually flip-flops between Page 1 and Page 2?

**Comment by jarodwsams** [comment]
That's not a bad idea... I can pull my unit from production and take it home to set up a test along those lines. Just let it go on and on until it fails.

**Comment by depili** [comment]
Timing based failures are really annoying to replicate, smashing few
buttons repeatedly so that there are many redraw actions cached sometimes
triggers one failed write, but haven't been able to see a single instance
of two failed writes in row.

**Comment by depili** [comment]
I have ran few tests where I absolutely swamp the rpi running companion
with state updates and so far I haven't been able to get the rate of write
failures to increase noticeably. The problem of generating multiple
updates for the pi is that the pi gets swamped quite fast in a weird way
as the events fill some kind of buffer and the deck updates will happen
really fast once the event flood stops... Directing the logging to
/dev/null helps a little on this.

One thing other than the modification of the companion code that I have
done is I purged the triggerhappy hid keyboard trigger daemon.

**Comment by depili** [comment]
So under heavy update load something strange happens, as the updates get
backed into some kind of buffer that is really deep (hundreds of updates)
and after the flood stops and little time passes the updates start getting
on the stream deck in really rapid pace eventually catching up to the real
time status. I had a similar problem earlier where the culprit was logging
to a fb console, but if I replace the hard coded debug filters with -*
this issue with companion is largely unaffected. (BTW, the DEBUG=-* set in
headless.js gets overwritten almost immediately by the one set in app.js,
and setting the defaults only if DEBUG env variable is not set by the
user would be nice).

Does generating new images have higher priorty than moving them towards
the streamdeck?
-- 
This message may contain traces of Truth.

**Comment by willosof** [comment]
I've cleaned up the debug stuff a bit now. Headless is default silent now, and when it comes to the normal "npm run start", it is now "yarn dev" (electron launcher + console debug) or "yarn prod"  (electron launcher, no debug)

**Comment by horuck** [comment]
I'm running a RPI 3B+ with streamdeck on 1.4.beta since today and experience this disconnection issue you are discussing above. 

Where do I need to change your fixing code and recompile then, so I get my Deck stable for now...?
Looking forward for your help.
Thanks.

**Comment by jarodwsams** [comment]
The file you're looking for is `/home/pi/companion/lib/usb/elgato.js`. The `draw` function starts on line 134. All you need to do is make the code modification and then reboot (`sudo reboot now`); no need to rebuild or anything like that.

> No guarantees this works. You're running 1.4-beta and this is being tested with 1.3-stable. It *should* work, but we don't even have this as an official fix for 1.3-stable.

**Comment by horuck** [comment]
thanks jarodwsams!
just did the change and will report back here later!

**Comment by jarodwsams** [comment]
If you still notice drops (I did, but only once when I was making **a lot** of changes all at once) try replacing my code block with the one from @Towerful. Mine tries it again just once after 20ms. His adds a retry counter (up to twice) after 20ms.

```javascript
elgato.prototype.draw = function(key, buffer, attempts = 0) {
        var self = this;
        if(!attempts){
                buffer = self.handleBuffer(buffer);
        }
        attempts++;

        try {
                self.streamDeck.fillImage(self.reverseButton(key), buffer);
                return true;
        } catch (e) {
                if(attempts > 2) {
                        self.log('Error drawing:', e);
                        self.system.emit('elgatodm_remove_device', self.devicepath);
                        return false;
                }
               setTimeout(self.draw.bind(self), 20, key, buffer, attempts)
                // alternatively a setImmediate() or nextTick()
        }
}
```

**Comment by horuck** [comment]
With your code replaced (with the redraw sectionI got it crashing again quite fast after a few minutes playing around with my hyperdeck mini, 
I then put in the code from @Towerful and all good so far... Will test further.

Thanks!

**Comment by jarodwsams** [comment]
If you haven't already, you should read the note at the top of the raspberrypi.md document.
https://github.com/bitfocus/companion/blob/master/documentation/raspberrypi.md

>Please note: performance on any Raspberry Pi system to date is less than optimal, and can easily break. **Running Companion in its current form on a Raspberry Pi is not recommended**. However, since the RPi is a popular device these instructions have been provided for you to use at your own risk. If you insist on running Companion on a Raspberry Pi, it is recommended to run Companion headless on the "Lite" version of the Raspbian OS. This will maximize the potential performance on your Raspberry Pi.

**Comment by horuck** [comment]
Code from @Towerful runs stable on my pi 3B+ with "regular" streamdeck so far.
@jarodwsams thanks for the not recommendation on pi. Will stick for a small linux pc for further installation and production use to have quicker USB reaction, http requests and emulator via Pi I find quite acceptable... Thanks for this great Software & Support!

**Comment by Coachwest2019** [comment]
I keep getting an error with line 54 anyone know what it could be and how to fix it?

**Comment by jarodwsams** [comment]
@Coachwest2019, we need some more info, and a screenshot or two.
What device?
What OS?
What OS Version?
What are you attempting to do when you get the error?
Can you share a screenshot of the error?

**Comment by Coachwest2019** [comment]
![error shot 1](https://user-images.githubusercontent.com/53186278/61742832-fc6dd480-ad61-11e9-8f6f-7c298944630e.PNG)
![error shot 2](https://user-images.githubusercontent.com/53186278/61742847-042d7900-ad62-11e9-9b3a-770fb1a8433a.PNG)
All on windows 10 this is for the companion

**Comment by jarodwsams** [comment]
@Coachwest2019 Since your issue is not related to the topic of this post, please create a new issue report. You can do that by clicking this link: [https://github.com/bitfocus/companion/issues/new/choose](https://github.com/bitfocus/companion/issues/new/choose). You'll be given a bug report template to fill out to make sure the community has all the information needed to assist you.

**Comment by willosof** [comment]
But, can we close this ticket? I'd say we have rpi headless support after headless.js :)

**Comment by jarodwsams** [comment]
> But, can we close this ticket? I'd say we've had rpi headless support after headless.js was introduced :)

I'd think so. Aside from the lag (known issue with the platform) it's been working quite swimmingly at church for several weeks now.

**Comment by depili** [comment]
I still suspect that there is a package installed in rasbian or a udev
rule or something like that which is the root cause of the write errors to
the streamdeck. We fixed that with the retry logic, but it would be nice
to find out what is causing the problem.

One idea would be to use raspian-ua-netinst to install a really minimal
environment and see if that also needs the retries. See
https://github.com/FooDeas/raspberrypi-ua-netinst

**Comment by jarodwsams** [comment]
@depili Honestly I wonder if the issue isn't so much what the software is doing, but the fact that the Ethernet runs over the USB bus on the RPi3. I've seen a lot of reports claiming that the RPi4 runs smooth as butter (at least on the 2GB version, unknown on the 1GB), and aside from an OS update the biggest relevant change is that the Ethernet was pulled off the USB bus.

I'm hoping to get a 4 4GB here soon and will be able to confirm the accuracy of those reports. I'd like to be able to get all 3 versions of the 4 (1GB, 2GB, 4GB RAM) to test side-by-side, but I don't think I'll be able to do that.

**Comment by josephdadams** [comment]
Housekeeping on the open issue here. We do have headless support... how is it working for you guys? Stable/recommended on pi4 only?

**Comment by alatteri** [comment]
working well on headless RPi4 for me.

**Comment by jarodwsams** [comment]
I would call it stable, with the caveat that you should not leave it always-on. Either shut down the Pi when it's not in use, or implement some method to have it automatically reboot on a regular schedule (I'd recommend daily). It is **strongly** recommended to not run Companion on a Pi 3, even a Pi 3 B+. Performance is terrible and laggy and not suitable for production. My personal recommendation is at minimum a Pi 4 2GB.

**Comment by haakonnessjoen** [comment]
Headless is working. The disconnection problem in RP3 should also be fixed now.

**Comment by jimyehoo** [comment]
I setup a pi 4 for pihole, then some other server function and companion.  The thing is, I need the pihole running most of the time while I would like to be able to remotely start Companion when I need it and shut it down when I am done.  Is there a mechanism to accomplish this?  a ssh script or something that I can create on Windows or Mac to remotely control it?  Companion is constantly trying the connect to the device instance and that's wasted bandwidth when none of the media activity is going on.  But wonderful job on the raspberry repository, loved it!

**Comment by josephdadams** [comment]
@jimyehoo there’s a new internal action in Companion to close it out. Assign that action to a button and then you can trigger that a number of ways. The easiest is probably curl/http. As far as opening Companion on schedule/trigger, it’s pretty easy to SSH and start a process.

**Comment by jimyehoo** [comment]
I actually have issues with Pi 4 install, the db copying says file is not found, it worked on Pi 2, but not on Pi 4.  I installed the Pi 4 several times and on many installs, it just doesn't work.  The install process doesn't report error but the db is always not there to be copied.  I am not sure if this is the cause, but I saw the compiled yarn is for armv7 and not armv8 which the Pi 4 cpu is on?  I have the Pi 4 doing lighting control and mumble chat, and the only thing I need on this Pi 4 is the Companion.

**Comment by jarodwsams** [comment]
![image](https://user-images.githubusercontent.com/22972834/65764623-b0396a80-e0eb-11e9-945b-08bdc564190f.png)

If you're talking about this error, I regularly see that as well. I had it when my production unit was a Pi 3 B+, as well, but I never saw any actual issues caused by it.

**Comment by jimyehoo** [comment]
yes, that warning comes up, too.  But I am talking during the installation process where if I just installed the Companion and run it the first time, I enter this command to copy the db... 
 cp ~/.config/companion/db ~/companion/
and it gives me error saying the file is not found.
cp: cannot stat '/home/pi/.config/companion/db': No such file or directory
but on some installation, companion runs after reboot, but on some installation, companion just not running.  I tried to use yarn prod to see if companion is running but it just complaint about can't find display

**Comment by jarodwsams** [comment]
@jimyehoo Ok, let's deal with that on an individual issue/bug report. Open a new report, with screenshots or log files if you can, with as much detail as possible, and tag me in it so I see it. I'll take a look at it as soon as I can.

New Issue/Bug Report: https://github.com/bitfocus/companion/issues/new/choose
